sway-lib-std/src/ecr.sw

```diff
@@ -1,31 +1,44 @@
 library ecr;
 
-use ::b512::B512;
 use ::address::Address;
+use ::b512::B512;
+use ::constants::ZERO;
+use ::hash::{HashMethod, hash_pair};
+use ::result::Result;
+
+pub enum EcRecoverError {
+    UnrecoverableAddress: (),
+    UnrecoverablePublicKey: (),
+}
 
 /// Recover the public key derived from the private key used to sign a message
-pub fn ec_recover(signature: B512, msg_hash: b256) -> B512 {
+pub fn ec_recover(signature: B512, msg_hash: b256) -> Result<B512, EcRecoverError> {
     let public_key = ~B512::new();
 
     asm(buffer: public_key.bytes, sig: signature.bytes, hash: msg_hash) {
         ecr buffer sig hash;
     };
-
-    public_key
+    /// if recovered key is empty
+    if (public_key.bytes)[0] == ZERO && (public_key.bytes)[1] == ZERO {
+        Result::Err(EcRecoverError::UnrecoverablePublicKey)
+    } else {
+        Result::Ok(public_key)
+    }
 }
 
 /// Recover the address derived from the private key used to sign a message
-pub fn ec_recover_address(signature: B512, msg_hash: b256) -> Address {
-    let address = asm(sig: signature.bytes, hash: msg_hash, addr_buffer, pub_key_buffer, hash_len: 64) {
-        move addr_buffer sp; // Buffer for address.
-        cfei i32;
-        move pub_key_buffer sp; // Temporary buffer for recovered key.
-        cfei i64;
-        ecr pub_key_buffer sig hash; // Recover public_key from sig & hash.
-        s256 addr_buffer pub_key_buffer hash_len; // Hash 64 bytes to the addr_buffer.
-        cfsi i64; // Free temporary key buffer.
-        addr_buffer: b256
-    };
+pub fn ec_recover_address(signature: B512, msg_hash: b256) -> Result<Address, EcRecoverError> {
+    let pub_key_result = ec_recover(signature, msg_hash);
 
-    ~Address::from(address)
+    if let Result::Ok(p) = pub_key_result {
+        let address = hash_pair((p.bytes)[0], (p.bytes)[1], HashMethod::Sha256);
+        if address != ZERO {
+            Result::Ok(~Address::from(address))
+        } else {
+            Result::Err(EcRecoverError::UnrecoverableAddress)
+        }
```

Comment:
Hashing should never error, so this block isn't needed. Just return the result immediately.

Reply:
yeah, I don't know what I was thinking here! The chances of having the hash be all zeros is not even worth considering.

Reply:
Ironically making the hash be zero is what PoW is all about... :slightly_smiling_face: Hrmm, is that ironic?  Bah.

Reply:
![external-content duckduckgo com](https://user-images.githubusercontent.com/3290375/162108492-f505929a-538d-4d1c-a286-a93e8bacef28.gif)